core/src/main/java/org/infinispan/functional/impl/AbstractFunctionalMap.java

```diff
@@ -114,20 +101,71 @@ private Transaction getOngoingTransaction() {
       }
    }
 
-   protected Object invoke(InvocationContext ctx, VisitableCommand cmd) {
-      Object result;
+   protected <T> CompletableFuture<T> invokeAsync(InvocationContext ctx, VisitableCommand cmd) {
+      CompletableFuture<T> cf;
+      boolean isImplicitTx = ctx.isInTxScope() && ((TxInvocationContext) ctx).isImplicitTransaction();
+      final Transaction implicitTransaction;
       try {
-         result = fmap.chain().invoke(ctx, cmd);
+         cf = (CompletableFuture<T>) fmap.chain.invokeAsync(ctx, cmd);
+         if (isImplicitTx) {
+            implicitTransaction = transactionManager.suspend();
```

Comment:
Same here, it's better to suspend the transaction before invoking the interceptors.